File: backend/app/domain/usecases/get_alumnos.py
from typing import List
from app.domain.repositories.alumno_repository import AlumnoRepository
from app.domain.entities.alumno import Alumno


class GetAlumnos:
    def __init__(self, repo: AlumnoRepository):
        self._repo = repo

    async def __call__(self) -> List[Alumno]:
        return await self._repo.list_all()

File: frontend/lib/domain/usecases/get_alumnos.dart
import '../entities/alumno.dart';
import '../repositories/alumno_repository.dart';

class GetAlumnos {
  const GetAlumnos(this.repository);
  final AlumnoRepository repository;

  Future<List<Alumno>> call({String? cursoId}) async {
    if (cursoId != null) {
      return repository.getAlumnosPorCurso(cursoId);
    }
    return repository.getAlumnos();
  }
}

class GetAlumnosEnRiesgo {
  const GetAlumnosEnRiesgo(this.repository);
  final AlumnoRepository repository;

  /// Retorna alumnos con asistencia < 75% (umbral RITE)
  Future<List<Alumno>> call() async {
    final todos = await repository.getAlumnos();
    return todos
        .where((a) => a.estadoRegularidad != EstadoRegularidad.regular)
        .toList()
      ..sort((a, b) => a.porcentajeAsistencia.compareTo(b.porcentajeAsistencia));
  }
}
